jobs: Rework Resumer API to eliminate the use of channels.

miretskiy opened the pull request:
Jobs infrastructure is responsible for running long jobs,
and restarting those jobs when needed (e.g. when the node goes down).
This infrastructure however, has been proven buggy and flaky, particularly
in the face of client connection termination, or the job being
adopted by another node.

Sometimes, the statements that create jobs (such as IMPORT) also
return data to the end user.  To this end, the resumer interface
accepted a channel to send the results to.

Unfortunately, the channel connected to the user session may disconnect.
Or, perhaps the node crashed and the job needed to be restarted eslewhere.
In such cases, we can never send the results to the user.  However,
there is significant complexity in the current implementation of the job
infrastructure, as well as the actual job implementations, that deals
with this problem.

StartableJob attempted to provides a mechanism to address some of that complexity
and allows the caller to start running the job and wait
for its completion via Run() method.  However, Run() method is
rather complex and error prone.

In addition, in order to support the functionality needed by the
Run method, namely the need to wait for the job completion,
the registry has to return the "error chanel" as a synchronization
mechanism, even though this error channel was used in very few places.

This PR reworks resumer API and the StartableJob itself to elliminate
relyance on channel.  Startable job, as its name implies, has
a Start method that simply starts running the job.

The callers may choose to wait for the completion, via AwaitCompletion
method, or they may choose not to wait, and allow the job run asynchrnously.

Waiting for startable job completion is simple: we only need to ensure that
the job running on this node registry finished without any errors.
If any errors were encountered when running the job, that error is returned
to the end user.  More importantly however, no matter why the error occured,
we never need to send row results back to the caller.

The caller of the startable job does have a channel to send the results to
the end user.  By waiting for the startable job to finish without error
on the gateway node, we can simply return whatever results the job produced.

In order to faciliate the jobs that need to return results, a JobResultReporter
interface is introduced. The jobs that do not produce
user visible results (e.g. auto stats update), simply do not implement
this interface.

Informs #54525

Release Notes: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/58939)
<!-- Reviewable:end -->

Commit:
jobs: Rework Resumer API to elliminate the use of channels.

Jobs infrastructure is responsible for running long jobs,
and restarting those jobs when needed (e.g. when the node goes down).
This infrastructure however, has been proven buggy and flaky, particularly
in the face of client connection termination, or the job being
adopted by another node.

Sometimes, the statements that create jobs (such as IMPORT) also
return data to the end user.  To this end, the resumer interface
accepted a channel to send the results to.

Unfortunately, the channel connected to the user session may disconnect.
Or, perhaps the node crashed and the job needed to be restarted eslewhere.
In such cases, we can never send the results to the user.  However,
there is significant complexity in the current implementation of the job
infrastructure, as well as the actual job implementations, that deals
with this problem.

StartableJob attempted to provides a mechanism to address some of that complexity
and allows the caller to start running the job and wait
for its completion via Run() method.  However, Run() method is
rather complex and error prone.

In addition, in order to support the functionality needed by the
Run method, namely the need to wait for the job completion,
the registry has to return the "error chanel" as a synchronization
mechanism, even though this error channel was used in very few places.

This PR reworks resumer API and the StartableJob itself to elliminate
relyance on channel.  Startable job, as its name implies, has
a Start method that simply starts running the job.

The callers may choose to wait for the completion, via AwaitCompletion
method, or they may choose not to wait, and allow the job run asynchrnously.

Waiting for startable job completion is simple: we only need to ensure that
the job running on this node registry finished without any errors.
If any errors were encountered when running the job, that error is returned
to the end user.  More importantly however, no matter why the error occured,
we never need to send row results back to the caller.

The caller of the startable job does have a channel to send the results to
the end user.  By waiting for the startable job to finish without error
on the gateway node, we can simply return whatever results the job produced.

In order to faciliate the jobs that need to return results, a JobResultReporter
interface is introduced. The jobs that do not produce
user visible results (e.g. auto stats update), simply do not implement
this interface.

Informs #54525

Release Notes: None

dt commented on pkg/ccl/changefeedccl/changefeed_stmt.go:
```diff
@@ -384,18 +379,10 @@ func changefeedPlanHook(
-			return ctx.Err()
-		case err := <-errCh:
-			return err
-		case <-startedCh:
```
so this is actually a change in behavior, right?

there's some code in `distChangefeedFlow` for which you dutifully make and pass a channel below, so that it can send on that ch in a little `finishedSetupFn` func that is passed to the distSQL planner to call after it starts up the flow. That func is sending on the ch, which prior to this change, was the "resultsCh" passed to the job, ie. startedCh above, passed in the usual resultCh position.


With that, then this recv would _block_ until the flow was setup (or one of the errors above), or until an error popped out of errCh. Without a call to await completion, we no longer see those execution errors that occur after creation but before startedCh would have unblocked, and without startedCh, we now return the job ID immediately, even if it will fail to start. Maybe that's okay, but it seems like there was intent here to avoid that.

We could try to use the ReportResults API instead here: put a ch on the changefeed resumer, have the `finishedSetupFn` callback func close it (along with a deferred close, in case it never makes it there), and then have `ReportResults` recv on it and then return nil (can ignore its resultsCh arg). Then a call to ReportResults right here would get us that same block-until-the-flow-is-up semantics. 

However we still aren't getting back exec errors, whereas I think we were if they occured before the flow it started and closed/sent on startedCh before. That'd be a little tricker, but I think we could in theory have Resume see if an error popped out of distflow, if so, store it in the resumer and then when we unblock on startedCh, we could return contents of that. Does that all make sense?

dt reviewed (approved):
LGTM except change in changefeed startup blocking/errors

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58939)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @miretskiy)

___
*[pkg/ccl/backupccl/backup_job.go, line 534 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58939#-MR1oN-YKX5ZxGPZ3dnP-r1-534:-MR_O1QgCynnwUsiGEyE:b-peyvc) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ccb85d48d2c4cfb69fd82937a08becd9382553/pkg/ccl/backupccl/backup_job.go#L534)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

do we need to select on context done too here, to avoid deadlock if we cancel execution after calling planhook.startExec but never call `Next()` to drain the ch?
</blockquote></details>

Yeah... I was thinking of pulling this select up to the caller, but that would work well if this method returned tree.Datums instead of taking a channel.

Select added.n
___
*[pkg/ccl/backupccl/restore_job.go, line 1314 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58939#-MR1oN-YKX5ZxGPZ3dnQ-r1-1314:-MR_OnTjDVc6aBOPcegQ:bs7wup3) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ccb85d48d2c4cfb69fd82937a08becd9382553/pkg/ccl/backupccl/restore_job.go#L1314)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

ditto on select
</blockquote></details>

Done.n
___
*[pkg/ccl/changefeedccl/changefeed_stmt.go, line 386 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58939#-MQw59OQ8NukQY3TnLgD-r1-386:-MR_P1tU6KmE9sre2OmO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ccb85d48d2c4cfb69fd82937a08becd9382553/pkg/ccl/changefeedccl/changefeed_stmt.go#L386)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

not new in this PR, but this seems like it needed to select on ctx done before -- was probably getting away with the fact it just did above and only seems once, but still in theory I think could deadlock without it.
</blockquote></details>

Done.
___
*[pkg/ccl/importccl/import_stmt.go, line 1237 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58939#-MQw59OQ8NukQY3TnLgE-r1-1237:-MR_PEFU9fPqQEh3eeuZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ccb85d48d2c4cfb69fd82937a08becd9382553/pkg/ccl/importccl/import_stmt.go#L1237)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

ditto again on select
</blockquote></details>

Done.
___
*[pkg/jobs/adopt.go, line 194 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58939#-MQw59OQ8NukQY3TnLgG-r1-194:-MR_PKvZ9fNXDdqIp0Y9:bki9y3u) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ccb85d48d2c4cfb69fd82937a08becd9382553/pkg/jobs/adopt.go#L194)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

why even bother putting the err into a chan here just to discard it below? I think that was originally done to then close resultsCh _after_ it was seems and discarded but with removal of resultsCh, I think errCh can go too and this can just be _ = r.runJob, removing the `go func() ...` on 199 too?
</blockquote></details>

You're right -- I was doing a mechanical replace here..
Makes sense.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58939)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @miretskiy)

___
*[pkg/jobs/jobs.go, line 889 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58939#-MQw59OQ8NukQY3TnLgH-r1-889:-MR_Qk973kh7Mwv1vrMR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ccb85d48d2c4cfb69fd82937a08becd9382553/pkg/jobs/jobs.go#L889)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I'd rather emit an error (or panic, since it is a serious developer error) than emit nothing silently?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> StartableJob attempted to provides a mechanism to address some of that complexity
> and allows the caller to start running the job and wait
> for its completion via Run() method.  However, Run() method is
> rather complex and error prone.


For history, startable job was only partially to address this complexity. It was more to deal with the fact that 

> Reviewed 3 of 25 files at r1, 5 of 15 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58939#-:-MRT248Z2kRZr8_CK3xj:b-lvcmdh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @miretskiy)

___
*[pkg/ccl/changefeedccl/changefeed_stmt.go, line 381 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58939#-MRa5gVR7paf5kklH742:-MRa5gVR7paf5kklH743:btbe86r) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccf43fa9e148320c7ffec47a1044ec95872e6b95/pkg/ccl/changefeedccl/changefeed_stmt.go#L381)):*
> ```Go
> 		}
> 
> 		// Start the job and wait for it to signal on startedCh.
> ```

stale comment
___
*[pkg/ccl/changefeedccl/changefeed_stmt.go, line 387 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58939#-MRaCCBY7K_iwQa4Fxqd:-MRaCCBZ8yDhmpth0xbM:bmulva2) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccf43fa9e148320c7ffec47a1044ec95872e6b95/pkg/ccl/changefeedccl/changefeed_stmt.go#L387)):*
> ```Go
> 
> 		select {
> 		case resultsCh <- tree.Datums{
> ```

This seems like a problematic and meaningful change in behavior. I'm worried about it. Before this change, if the changefeed ran into an error before a certain point in startup, that error would make it to the user.


<!-- Sent from Reviewable.io -->

dt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58939)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @miretskiy)

___
*[pkg/ccl/backupccl/backup_job.go, line 534 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58939#-MR1oN-YKX5ZxGPZ3dnP-r1-534:-MR_SonI75YYt7YaiNgE:bvpb3b0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ccb85d48d2c4cfb69fd82937a08becd9382553/pkg/ccl/backupccl/backup_job.go#L534)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Yeah... I was thinking of pulling this select up to the caller, but that would work well if this method returned tree.Datums instead of taking a channel.

Select added.n
</blockquote></details>

minor nit: we usually select on context first in case they're both runnable, but doesn't really matter here.
___
*[pkg/ccl/changefeedccl/changefeed_stmt.go, line 387 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58939#-MRaCCBY7K_iwQa4Fxqd:-MRaWrli8QSp7KZsosyw:b-sf617o) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccf43fa9e148320c7ffec47a1044ec95872e6b95/pkg/ccl/changefeedccl/changefeed_stmt.go#L387)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

This seems like a problematic and meaningful change in behavior. I'm worried about it. Before this change, if the changefeed ran into an error before a certain point in startup, that error would make it to the user.
</blockquote></details>

I made a similar comment but reviewable ate it? https://github.com/cockroachdb/cockroach/pull/58939#discussion_r561630143


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58939)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @dt)

___
*[pkg/ccl/backupccl/backup_job.go, line 534 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58939#-MR1oN-YKX5ZxGPZ3dnP-r1-534:-MRanJG39dp4hx3Lt74y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ccb85d48d2c4cfb69fd82937a08becd9382553/pkg/ccl/backupccl/backup_job.go#L534)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

minor nit: we usually select on context first in case they're both runnable, but doesn't really matter here.
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/changefeed_stmt.go, line 387 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58939#-MRaCCBY7K_iwQa4Fxqd:-MRanmTs3FbhLp_MTFNG:buub24z) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccf43fa9e148320c7ffec47a1044ec95872e6b95/pkg/ccl/changefeedccl/changefeed_stmt.go#L387)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I made a similar comment but reviewable ate it? https://github.com/cockroachdb/cockroach/pull/58939#discussion_r561630143
</blockquote></details>

Spoke w/ @ajwerner 

I didn't want to abuse the channel mechanism this way.  Turns out, we don't really give the user anymore guarantees, whether we wait or not.  AN error could happen immediately after we invoke callback function, but before we start running.   `distChangefeedFlow` method does not give guarnatees for example that channel is signaled iff there is no error -- it's quite possible for this method to signal the channel, and still return an error (i.e. we return job ID to the user sahing "all is good" when in reality not all is good).

In short, @ajwerner  was concerned that we may report large number of issues late (thus making UX worse).  However, `validateDetails` function does majority of "silly mistakes" checking already -- and that's something that we do already during the planning.

Therefore, I think it makes sense to just start the change feed job, without adding workarounds w/ channels  (particularly since those workarounds are not that good anyway)
___
*[pkg/jobs/deprecated.go, line 273 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58939#-MR1oN-YKX5ZxGPZ3dnU-r1-273:-MRaHu014O747gn0S8W-:b-6ktr4c) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ccb85d48d2c4cfb69fd82937a08becd9382553/pkg/jobs/deprecated.go#L273)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

ditto, we're not (usefully) reading from errCh anymore, right?
</blockquote></details>

I didn't want to get into fixing something that's marked deprecated...


<!-- Sent from Reviewable.io -->

dt reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58939)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @miretskiy)

___
*[pkg/jobs/deprecated.go, line 273 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58939#-MR1oN-YKX5ZxGPZ3dnU-r1-273:-MRb78IXDi21TZ0CCvjY:bsod8fb) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ccb85d48d2c4cfb69fd82937a08becd9382553/pkg/jobs/deprecated.go#L273)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I didn't want to get into fixing something that's marked deprecated...
</blockquote></details>

eh, sure, but I think it's actually worse to leave it as is, because this is the only callsite: the function makes a chan to return and writes one value, all so we can spin up a useless goroutine to discard that value. deprecated or not, I think it'd be less confusing to delete it outright.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 6 of 25 files at r1, 6 of 15 files at r2, 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58939#-:-MRb95lb3uRhgo4HfwHm:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dt and @miretskiy)



<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58939)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dt)

___
*[pkg/jobs/deprecated.go, line 273 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58939#-MR1oN-YKX5ZxGPZ3dnU-r1-273:-MRbahXo4RUXSi1IvOoz:b-ql3zu) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ccb85d48d2c4cfb69fd82937a08becd9382553/pkg/jobs/deprecated.go#L273)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

eh, sure, but I think it's actually worse to leave it as is, because this is the only callsite: the function makes a chan to return and writes one value, all so we can spin up a useless goroutine to discard that value. deprecated or not, I think it'd be less confusing to delete it outright.
</blockquote></details>

It's not quite useless -- this channel is used to synchronize the call site with job completion.
I use channel in startable job for the same purpose -- it's just this deprecated code path does not use startable job.


<!-- Sent from Reviewable.io -->

dt reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58939)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dt and @miretskiy)

___
*[pkg/jobs/deprecated.go, line 273 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58939#-MR1oN-YKX5ZxGPZ3dnU-r1-273:-MReahza0km3drDYah4X:b-kieoi6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0ccb85d48d2c4cfb69fd82937a08becd9382553/pkg/jobs/deprecated.go#L273)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

It's not quite useless -- this channel is used to synchronize the call site with job completion.
I use channel in startable job for the same purpose -- it's just this deprecated code path does not use startable job.
</blockquote></details>

but this is the only call to `r.deprecatedResume`, so it allocates `errCh` and sends on it all so we can start a new goroutine below to drain it discarding the value, without looking at it or doing anything after draining it. Prior to this change, draining it did synchronize closing the resultsCh, but after this change that ch is gone so I'm pretty sure it is now doing nothing, just adding a little confusion/some extra cpu cycles and mutex locks. Am I missing something?

I'm fine merging as-is though -- if I am correct in that reading, I can just delete the extra code in a pure-delete PR following this one.


<!-- Sent from Reviewable.io -->

miretskiy commented:
bors r+

miretskiy commented:
bors r+
